### ISSUE TITLE
개정 대상인 불변 개체의 속성 이름 제약사항

### ISSUE THREAD

**Issue by wplong11** [AutoReviser/AutoReviser#2]
# 불변 개체의 속성 이름 규칙 제안
드물게 발생할거라고 생각하지만, 생성자 매개변수 또는 속성의 이름이 중복된 경우 개체를 개정할 때 발생하는 문제를 고민합니다.

## 문제

아래 예제 코드와 같이 `StringComparison.OrdinalIgnoreCase` 조건에서 중복된 이름을 가진 경우, `Revise` 메소드의 작동이 사용자의 기대와 다를 수 있습니다.
```C#
class IgnoreCaseImmutableObject
{
    public IgnoreCaseImmutableObject(int foo, int foO)
    {
        (Foo, FoO) =  (foo, foO);
    }

    public int Foo { get; }

    public int FoO { get; }
}

// Arrange
var seed = new IgnoreCaseImmutableObject(1, 2);

// Act
IgnoreCaseImmutableObject actual = seed.Revise(x => x.Foo == 3);

// Assert
Assert.AreEqual(3, actual.Foo);
Assert.AreEqual(seed.FoO, actual.FoO); // actual.FoO 도 3으로 변경되는 문제 발생.
```

554b2c58d0a94fbb849ad69e453a633086c0d1d2 커밋을 기준으로 아래 두개 메소드를 살펴보면 문제 원인을 이해할 수 있습니다.
1. 대/소문자를 무시해서 생성자 매개변수 이름과 속성 이름을 비교
https://github.com/AutoReviser/AutoReviser/blob/554b2c58d0a94fbb849ad69e453a633086c0d1d2/source/AutoReviser/ConstructorArgument.cs#L23-L25

2. 동일한 이름을 가진 모든 생성자 매개변수의 값을 변경
https://github.com/AutoReviser/AutoReviser/blob/554b2c58d0a94fbb849ad69e453a633086c0d1d2/source/AutoReviser/ConstructorInvoker.cs#L49-L59

## 제안
동일한 속성 이름을 가진 개체를 개정하는 로직은 사용 빈도가 낮아보이고, 이 경우를 제대로 지원하는 논리는 꽤나 모호하고 어려워보입니다. 이 문제를 당장 해결할 계획이 아니라면 우선 동일한 이름을 가진 불변 개체를 개정하는 것을 막는 것은 어떨까요? 해당 조건의 불변 개체를 개정 시도하면 `NotSupportedException` 를 던지게 처리하고요.

```C#
        [TestMethod]
        [AutoData]
        public void Revise_throw_not_supported_exception_when_when_the_name_of_constructor_parameter_is_duplicated(
            IgnoreCaseImmutableObject seed)
        {
            // Act
            IgnoreCaseImmutableObject actual = null;
            Action revise = () =>
            {
                actual = seed.Revise(x => x.Foo == 3);
            };

            // Assert
            revise.Should().Throw<NotSupportedException>();
        }

        [TestMethod]
        [AutoData]
        public void Revise_throw_not_supported_exception_when_when_the_name_of_property_is_duplicated(
            IgnoreCaseImmutableObject seed)
        {
            // Act
            IgnoreCaseImmutableObject actual = null;
            Action revise = () =>
            {
                 actual = seed.Revise(x => x.Foo == 3);
            };

            // Assert
            revise.Should().Throw<NotSupportedException>();
        }
```

**Comment by gyuwon** [comment]
매개변수 이름을 속성 이름과 비교할 때 속성 이름에서 첫 글자만 소문자이고 다른 글자는 정확히 일치하는 경우만 같은 것으로 판단하게 하면 어떨까 싶네요. 만약 속성 이름의 첫 글자가 영문 대문자가 아니라면 모든 글자가 일치하는 경우로 하고요.

|속성 이름|매개변수 이름|일치여부|
|--|--|--|
|`Foo`|`foO`|**false**|
|`Foo`|`foo`|**true**|
|`@Foo`|`@foo`|**false**|
|`@Foo`|`@Foo`|**true**|

**Comment by wplong11** [comment]
제안하신 판단 논리 사용하면 다음과 같은 속성을 가지는 경우도 대응할 수 있겠네요.
```c#
class IgnoreCaseImmutableObject
{

    ...

    public int Foo { get; }

    public int foo { get; }
}
```

**Comment by gyuwon** [comment]
그런 경우는 생각하지 못했는데, 이 땐 제가 제시한 논리는 문제를 일으킬 것 같네요. `Foo` 속성도 `foo` 속성도 `foo` 매개변수를 찾게되니까요.

**Comment by wplong11** [comment]
일단 제시해주신 [규칙](https://github.com/AutoReviser/AutoReviser/issues/2#issuecomment-405025770)에서 문제가 발생하지 않는다고 생각한 이유는 애초에 해당 규칙을 만족하는 생성자를 선언할 수 없다고 생각했기 때문입니다. 넘겨받은 인자를 단순히 속성에 대입시키는게 생성자 역할의 전부라면, `foo`, `Foo` 두개 속성에 대응되는 매개변수 이름이 충돌되니까 생성자 선언이 불가능하다고 생각했습니다. 그런데 말씀하신 [문제](https://github.com/AutoReviser/AutoReviser/issues/2#issuecomment-405027461)가 다음과 같은 경우에 발생할 수 있겠네요. 🤔

```c#
class IgnoreCaseImmutableObject
{
    public IgnoreCaseImmutableObject(int foo)
    {
        this.foo = foo;
        this.Foo = foo * 2;
    }

    public int foo { get; }

    public int Foo { get; }
}
```

**Comment by gyuwon** [comment]
`ConstructorResolver`가 우선 공용 속성의 수와 동일한 수의 매개변수를 가지는 생성자를 찾도록 하는 것이 어떨까 싶네요.

**Comment by wplong11** [comment]
> ConstructorResolver가 우선 공용 속성의 수와 동일한 수의 매개변수를 가지는 생성자를 찾도록 하는 것이 어떨까 싶네요.

`우선` 이라는 말은 공용 속성 수와 동일한 수의 매개변수를 가지는 생성자를 찾는 것을 먼저 시도하고, 실패한 경우 다른 생성자를 찾는 논리를 의미하나요?

**Comment by gyuwon** [comment]
우선 해당 검사 코드 부터 먼저 추가하는 것이 어떨까 하는 의미에요. '우선'이 맨 앞에 나왔으면 의미 전달이 더 정확했을 것 같네요. :-)

**Comment by wplong11** [comment]
**ConstructorResolver** 가 생성자 공용 속성과 매개변수의 수를 비교하는 것과 더불어 이름 그리고 타입까지 비교하는 것은 어떨까요?

fe15b4ab01e6d0040c15967153691db502ad24d9 기준으로 수 비교 논리만 추가하면, 다음과 같은 경우에도 `Revise` 작동에 문제가 발생할 수 있습니다.

- 공용 속성과 동일한 수의 매개변수를 가진 생성자가 단수 개인 경우
    - 공용 속성과 매개변수의 수는 같지만 다른 이름이 존재하는 경우
      ```C#
      public IgnoreCaseImmutableObject(int foo, int bar)
      {
          this.foo = foo;
          this.Foo = bar;
      }
      ```
    - 공용 속성과 매개변수의 이름 집합은 동일하지만 다른 타입이 존재하는 경우
      ```C#

      public IgnoreCaseImmutableObject(int foo, string bar)
      {
          this.Foo = foo;
          this.Bar = int.Parse(bar);
      }

      public int Foo { get; }

      public int Bar { get; }
      ```

-  공용 속성과 동일한 수의 매개변수를 가진 생성자가 복수 개 있는 경우
    - 공용 속성과 매개변수의 수는 같지만 다른 이름이 존재하는 경우
    - 공용 속성과 매개변수의 이름 집합은 동일하지만 다른 타입이 존재하는 경우

**Comment by gyuwon** [comment]
맞아요. 제가 '우선'이란 표현을 사용한 것은 `ConstructorResolver`에 가장 명확한 기준 하나를 먼저 추가하는 것이 어떻겠냐는 것입니다. 지금 이슈 #3 을 처리하면서도 생성자가 아닌 공장 함수를 가진 클래스에 대해서는 `ConstructorResolver` 설계를 어떻게 수정해야 할 것인가 고민하는 중이라 섣불리 설계를 확장하기 쉽지 않네요.
